### PR TITLE
rustdoc: Replace String with Symbol in `Cache.paths` et al.

### PR DESCRIPTION
**Pull request by camelid**
**Blocked on #91871.**

This should hopefully improve memory usage and rustdoc run time.
It's also more consistent with the rest of rustc and rustdoc.

r? @jyn514

### CODE REVIEW

File: src/librustdoc/html/url_parts_builder.rs [hunk 1]
```diff
@@ -115,5 +117,22 @@ impl<'a> Extend<&'a str> for UrlPartsBuilder {
...
+    }
+}
+
+impl Extend<Symbol> for UrlPartsBuilder {
+    fn extend<T: IntoIterator<Item = Symbol>>(&mut self, iter: T) {
+        let iter = iter.into_iter();
+        self.buf.reserve(AVG_PART_LENGTH * iter.size_hint().0);
+        iter.for_each(|part| self.push(&part.as_str()));
+    }
+}
```
Comment (camelid) [inline src/librustdoc/html/url_parts_builder.rs:135]: I'm worried this is too "magical", but it does really help at several call sites.

**Comment by camelid** [comment]
Hmm, somehow this broke linking to external items.

**Comment by camelid** [comment]
#91948 covers this and its CI is passing. Closing this PR.

**Comment by nnethercote** [comment]
Oh geez, my first ever contribution to rustdoc and I'm stepping on your toes... sorry :(  Unfortunate timing.

**Comment by camelid** [comment]
> Oh geez, my first ever contribution to rustdoc and I'm stepping on your toes... sorry :( Unfortunate timing.

Thanks ❤️

At least I'll be able to review your PR better ;)

**Pull request closed without merging**